File: IsraelHiking.Web/android/app/src/main/java/com/mapeak/valhalla/ValhallaProfiles.kt
package com.mapeak.valhalla

import android.content.Context
import org.json.JSONObject
import java.io.File

/**
 * A single routing profile: the valhalla costing model and the costing options to use with it, as
 * the json they were stored as, so that they can be read into the models of that costing model.
 */
data class ValhallaProfile(val costing: String, val costingOptionsJson: String?)

/**
 * The routing profiles on the device.
 *
 * They are handed over after being downloaded with the offline maps, and are the same profiles the
 * server routes with, so that a route on the device follows the same costing options as one online.
 * Without them there is no offline routing, the same as without tiles.
 */
class ValhallaProfiles(private val context: Context) {

    companion object {
        /** The plugin's own copy, the app hands it the file it downloaded */
        private const val PROFILES_FILE_NAME = "valhalla_profiles.json"
        private const val COSTING_KEY = "costing"
        private const val COSTING_OPTIONS_KEY = "costingOptions"
    }

    private fun profilesFile(): File = File(context.filesDir, PROFILES_FILE_NAME)

    /**
     * Keeps the profiles as they were downloaded, they are only read when a route is calculated.
     */
    fun store(profiles: String) {
        // Fail here rather than when routing, so that content that can not be read is never kept
        JSONObject(profiles)
        profilesFile().writeText(profiles)
    }

    /**
     * The profile of the given name, or null when the profiles were not downloaded or hold no such
     * profile.
     */
    fun get(name: String): ValhallaProfile? {
        val file = profilesFile()
        if (!file.exists()) {
            return null
        }
        val profile = JSONObject(file.readText()).optJSONObject(name) ?: return null
        val costing = profile.optString(COSTING_KEY)
        if (costing.isEmpty()) {
            return null
        }
        return ValhallaProfile(costing, profile.optJSONObject(COSTING_OPTIONS_KEY)?.toString())
    }
}

File: IsraelHiking.Web/ios/App/App/Valhalla/ValhallaProfiles.swift
import Foundation

/**
 * A single routing profile: the valhalla costing model and the costing options to use with it.
 */
enum ValhallaProfilesError: Error {
    case invalidProfiles
}

struct ValhallaProfile {
    let costing: String
    let costingOptions: [String: Any]?
}

/**
 * The routing profiles on the device, mirroring `ValhallaProfiles.kt`.
 *
 * They are handed over after being downloaded with the offline maps, and are the same profiles the
 * server routes with, so that a route on the device follows the same costing options as one online.
 * Without them there is no offline routing, the same as without tiles.
 */
final class ValhallaProfiles {
    /// The plugin's own copy, the app hands it the file it downloaded
    private static let profilesFileName = "valhalla_profiles.json"
    private static let costingKey = "costing"
    private static let costingOptionsKey = "costingOptions"

    private let fileManager = FileManager.default

    private var profilesURL: URL {
        fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent(ValhallaProfiles.profilesFileName)
    }

    /**
     * Keeps the profiles as they were downloaded, they are only read when a route is calculated.
     */
    func store(_ profiles: String) throws {
        // Fail here rather than when routing, so that content that can not be read is never kept
        guard let data = profiles.data(using: .utf8),
              (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] != nil else {
            throw ValhallaProfilesError.invalidProfiles
        }
        try data.write(to: profilesURL, options: .atomic)
    }

    /**
     * The profile of the given name, or nil when there are no profiles or no such profile in them.
     */
    func profile(named name: String) -> ValhallaProfile? {
        guard let data = try? Data(contentsOf: profilesURL),
              let profiles = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let profile = profiles[name] as? [String: Any],
              let costing = profile[ValhallaProfiles.costingKey] as? String, !costing.isEmpty else {
            return nil
        }
        return ValhallaProfile(costing: costing, costingOptions: profile[ValhallaProfiles.costingOptionsKey] as? [String: Any])
    }
}
